File: jp.shiranui-isuzu.unity-mcp/Editor/Handlers/CompileCommandHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEditor.Compilation;
using UnityEngine;
using UnityMCP.Editor.Core;

namespace UnityMCP.Editor.Handlers
{
    [InitializeOnLoad]
    internal sealed class CompileCommandHandler : IMcpCommandHandler
    {
        private static readonly List<string> LastErrors = new List<string>();
        private static readonly List<string> LastWarnings = new List<string>();
        private static bool _compilationEverFinished = false;

        static CompileCommandHandler()
        {
            CompilationPipeline.compilationStarted += _ =>
            {
                lock (LastErrors)
                {
                    LastErrors.Clear();
                    LastWarnings.Clear();
                }
                _compilationEverFinished = false;
            };

            CompilationPipeline.assemblyCompilationFinished += (path, messages) =>
            {
                lock (LastErrors)
                {
                    foreach (var msg in messages)
                    {
                        var entry = $"{msg.file}({msg.line},{msg.column}): {msg.message}";
                        if (msg.type == CompilerMessageType.Error)
                            LastErrors.Add(entry);
                        else if (msg.type == CompilerMessageType.Warning)
                            LastWarnings.Add(entry);
                    }
                }
            };

            CompilationPipeline.compilationFinished += _ => { _compilationEverFinished = true; };
        }

        public string CommandPrefix => "compile";
        public string Description => "Trigger and monitor Unity script compilation";
        public McpIdempotency Idempotency => McpIdempotency.Unsafe;

        public IReadOnlyList<McpHandlerAction> Actions { get; } = new[]
        {
            new McpHandlerAction("refresh",   McpIdempotency.Unsafe),
            new McpHandlerAction("getStatus", McpIdempotency.Safe),
        };

        public JObject Execute(string action, JObject parameters)
        {
            return action.ToLower() switch
            {
                "refresh"   => TriggerRefresh(),
                "getstatus" => GetStatus(),
                _ => new JObject
                {
                    ["error"] = $"Unknown action: {action}. Supported: refresh, getStatus"
                }
            };
        }

        private static JObject TriggerRefresh()
        {
            AssetDatabase.Refresh();
            return new JObject
            {
                ["triggered"] = true,
                ["isCompiling"] = EditorApplication.isCompiling,
                ["message"] = "AssetDatabase.Refresh() called. If .cs files changed, compilation will begin (brief disconnection during domain reload is normal)."
            };
        }

        private static JObject GetStatus()
        {
            var isCompiling = EditorApplication.isCompiling;
            string[] errors, warnings;
            lock (LastErrors)
            {
                errors = LastErrors.ToArray();
                warnings = LastWarnings.ToArray();
            }

            var status = isCompiling ? "compiling" :
                         errors.Length > 0 ? "errors" : "ready";

            return new JObject
            {
                ["isCompiling"] = isCompiling,
                ["status"] = status,
                ["errorCount"] = errors.Length,
                ["warningCount"] = warnings.Length,
                ["errors"] = new JArray(errors.Cast<object>()),
                ["warnings"] = new JArray(warnings.Cast<object>()),
                ["message"] = status switch
                {
                    "compiling" => "Unity is currently compiling scripts. Poll again until isCompiling=false.",
                    "errors" => $"Compilation finished with {errors.Length} error(s). Fix the errors before running code.",
                    _ => _compilationEverFinished
                        ? "Scripts compiled successfully. Ready to execute."
                        : "No compilation has occurred yet in this session."
                }
            };
        }
    }
}

File: unity-mcp-ts/src/handlers/CompileCommandHandler.ts
import { IMcpToolDefinition } from "../core/interfaces/ICommandHandler.js";
import { JObject } from "../types/index.js";
import { z } from "zod";
import { BaseCommandHandler } from "../core/BaseCommandHandler.js";

const DEFAULT_TIMEOUT_MS = 30_000;
const POLL_INTERVAL_MS = 500;
const INITIAL_WAIT_MS = 800;

/**
 * Provides two MCP tools for Unity script compilation:
 *
 * - compile_scripts: triggers AssetDatabase.Refresh() and waits (polling) until
 *   compilation completes, then returns status + any errors. Call this after
 *   creating or editing .cs project files.
 *
 * - compile_status: read-only status check — returns current isCompiling state
 *   and error list without triggering a recompile.
 */
export class CompileCommandHandler extends BaseCommandHandler {
    public get commandPrefix(): string {
        return "compile";
    }

    public get description(): string {
        return "Trigger and monitor Unity script compilation";
    }

    protected async executeCommand(action: string, parameters: JObject): Promise<JObject> {
        switch (action.toLowerCase()) {
            case "scripts":
                return this.compileAndWait(parameters);
            case "status":
                return this.getStatus(parameters);
            default:
                return {
                    success: false,
                    error: `Unknown action: ${action}. Supported actions: scripts, status`
                };
        }
    }

    public getToolDefinitions(): Map<string, IMcpToolDefinition> {
        const tools = new Map<string, IMcpToolDefinition>();

        tools.set("compile_scripts", {
            description:
                "Trigger Unity script recompilation and wait until it finishes. " +
                "IMPORTANT: Call this after creating or editing any .cs project files " +
                "before calling unity_execute_code or menu_execute. " +
                "Returns: status ('ready'|'errors'), errorCount, errors[], warningCount. " +
                "A domain reload may occur during compilation — this is normal.",
            parameterSchema: {
                timeoutMs: z
                    .number()
                    .optional()
                    .describe(`Maximum wait time in milliseconds (default: ${DEFAULT_TIMEOUT_MS})`)
            },
            annotations: {
                title: "Compile Unity Scripts",
                readOnlyHint: false,
                destructiveHint: false,
                idempotentHint: true,
                openWorldHint: false
            }
        });

        tools.set("compile_status", {
            description:
                "Get current Unity compilation status without triggering a recompile. " +
                "Returns: isCompiling (bool), status ('compiling'|'errors'|'ready'), " +
                "errorCount, errors[], warningCount, message.",
            parameterSchema: {},
            annotations: {
                title: "Get Compilation Status",
                readOnlyHint: true,
                idempotentHint: true,
                openWorldHint: false
            }
        });

        return tools;
    }

    private async compileAndWait(parameters: JObject): Promise<JObject> {
        const timeoutMs = (parameters.timeoutMs as number) ?? DEFAULT_TIMEOUT_MS;

        // Trigger AssetDatabase.Refresh on the Unity side.
        // If scripts changed, Unity will start a domain reload — the connection
        // may drop briefly, which is expected and handled by the retry logic.
        try {
            await this.sendUnityRequest("compile.refresh", {});
        } catch {
            // Domain reload in progress — keep going
        }

        // Give Unity a moment to start compiling before the first poll.
        await sleep(INITIAL_WAIT_MS);

        const deadline = Date.now() + timeoutMs;

        while (Date.now() < deadline) {
            try {
                const status = await this.sendUnityRequest("compile.getStatus", {});
                if (status.isCompiling === false) {
                    return status;
                }
            } catch {
                // Unity still reloading — wait and retry
            }
            await sleep(POLL_INTERVAL_MS);
        }

        return {
            status: "timeout",
            isCompiling: true,
            errorCount: 0,
            warningCount: 0,
            errors: [],
            warnings: [],
            message: `Compilation did not finish within ${timeoutMs}ms. Check Unity Editor for details.`
        };
    }

    private async getStatus(parameters: JObject): Promise<JObject> {
        return this.sendUnityRequest("compile.getStatus", parameters);
    }
}

function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
}
